pay-api/src/pay_api/services/eft_refund.py

```diff
@@ -0,0 +1,272 @@
+                                invoice_id=invoice.id,
+                                is_processing=True,
+                                hidden=False)).flush()
+
+        return InvoiceStatus.REFUND_REQUESTED.value
+
+    @staticmethod
+    def reverse_eft_credits(shortname_id: int, amount: Decimal):
+        """Reverse the amount to eft_credits table based on short_name_id."""
+        eft_credits = EFTCreditModel.get_eft_credits(shortname_id, include_zero_remaining=True)
+        for credit in eft_credits:
+            if credit.remaining_amount == credit.amount:
+                continue
+            if credit.remaining_amount > credit.amount:
+                raise BusinessException(Error.EFT_CREDIT_AMOUNT_UNEXPECTED)
+
+            credit_adjustment = min(amount, credit.amount - credit.remaining_amount)
+            amount -= credit_adjustment
+            credit.remaining_amount += credit_adjustment
+            credit.flush()
+
+        # Scenario where we tried to reverse an amount all the credits,
+        # but for some reason our refund was more than the original amount on the credits.
+        if amount > 0:
+            raise BusinessException(Error.INVALID_REFUND)
+
+    @staticmethod
+    def refund_eft_credits(shortname_id: int, amount: Decimal):
+        """Refund the amount to eft_credits table based on short_name_id."""
+        eft_credits = EFTCreditModel.get_eft_credits(shortname_id)
+        eft_credit_balance = EFTCreditModel.get_eft_credit_balance(shortname_id)
+        if amount > eft_credit_balance or amount <= 0:
+            raise BusinessException(Error.INVALID_REFUND)
+
+        for credit in eft_credits:
+            credit_amount = Decimal(credit.remaining_amount)
+            if credit_amount <= 0:
+                continue
+
+            deduction = min(amount, credit_amount)
+            credit.remaining_amount -= deduction
+            amount -= deduction
+
+            credit.flush()
+        # Scenario where we couldn't subtract the remaining amount from the credits.
+        if amount > 0:
+            raise BusinessException(Error.INVALID_REFUND)
+
+    @staticmethod
+    def update_shortname_refund(refund_id: int, data: EFTShortNameRefundPatchRequest) -> EFTRefundModel:
+        """Update the refund status."""
+        refund = EFTRefundModel.find_by_id(refund_id)
+        if refund.status != EFTShortnameRefundStatus.PENDING_APPROVAL.value:
+            raise BusinessException(Error.REFUND_ALREADY_FINALIZED)
+        refund.comment = data.comment
+        refund.status = data.status
+        refund.decline_reason = data.decline_reason
+        refund.save_or_add(auto_save=False)
+        short_name = EFTShortnamesModel.find_by_id(refund.short_name_id)
+        match data.status:
+            case EFTShortnameRefundStatus.DECLINED.value:
+                EFTRefund.reverse_eft_credits(refund.short_name_id, refund.refund_amount)
+                EFTHistoryService.create_shortname_refund(
```

Comment:
For both declined and approved we can find the existing history record by eft_refund_id. 

For:
1) Approved - transition the history transaction_type to `EFTHistoricalTypes.SN_REFUND_APPROVED.value`

2) Declined - transition the history transaction_type to `EFTHistoricalTypes.SN_REFUND_DECLINED.value` and update the credit balance to the latest `EFTCreditModel.get_eft_credit_balance(refund.short_name_id)`